crates/core/tedge_actors/src/servers/requests.rs

```diff
@@ -0,0 +1,229 @@
+use crate::Actor;
+use crate::ChannelError;
+use crate::DynSender;
+use crate::LoggingReceiver;
+use crate::Message;
+use crate::MessageReceiver;
+use crate::RuntimeError;
+use crate::RuntimeRequest;
+use crate::Sender;
+use crate::Server;
+use async_trait::async_trait;
+use futures::channel::oneshot;
+use futures::StreamExt;
+use log::error;
+use std::fmt::Debug;
+use std::ops::ControlFlow;
+
+/// Wrap a request with a [Sender] to send the response to
+///
+/// Requests are sent to server actors using such envelopes telling where to send the responses.
+pub struct RequestEnvelope<Request, Response> {
+    pub request: Request,
+    pub reply_to: Box<dyn Sender<Response>>,
+}
+
+impl<Request: Debug, Response> Debug for RequestEnvelope<Request, Response> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.request.fmt(f)
+    }
+}
+
+/// A message box used by a client to request a server and await the responses.
+pub struct ClientMessageBox<Request, Response> {
+    sender: DynSender<RequestEnvelope<Request, Response>>,
+}
+
+impl<Request: Message, Response: Message> ClientMessageBox<Request, Response> {
+    /// Send the request and await for a response
+    pub async fn await_response(&mut self, request: Request) -> Result<Response, ChannelError> {
+        let (sender, receiver) = oneshot::channel::<Response>();
+        let reply_to = Box::new(Some(sender));
+        self.sender
+            .send(RequestEnvelope { request, reply_to })
+            .await?;
+        let response = receiver.await;
+        response.map_err(|_| ChannelError::ReceiveError())
+    }
+}
+
+/// A [Sender] used by a client to send requests to a server,
+/// redirecting the responses to another recipient.
+pub struct RequestSender<Request: 'static, Response: 'static> {
```

Comment:
This same struct seems to be defined in `message_boxes.rs` as well. What's the difference?

Reply:
My bad: `request.rs` is a draft version I forgot to remove.

Reply:
`request.rs` has been removed by https://github.com/thin-edge/thin-edge.io/pull/2600/commits/2556bb40bee98e9d5a7a8c57d49d6800afb5a549